drain-service/drain_training_inferencing.py

```diff
@@ -33,46 +29,147 @@
+        logging.info(workload_template_miner.drain.clusters_counter)
+        for log_data in logs_payload.items:
+            log_message = log_data.masked_log
+            if log_message:
+                current_template_matched, current_template_payload = match_template(log_data, workload_template_miner)
+                if current_template_matched:
+                    logs_inferenced_results.append(log_data)
+                    if current_template_payload:
+                        log_templates_modified.append(current_template_payload)
+                else:
+                    workload_model_logs.append(log_data)
+        if (start_time - last_time >= 1) or len(logs_inferenced_results) >= 128 or len(log_templates_modified) >= 128:
+            if len(logs_inferenced_results) > 0:
+                await nw.publish("inferenced_logs", bytes(PayloadList(items=logs_inferenced_results)))
+                logs_inferenced_results = []
+            if len(log_templates_modified) > 0:
+                await nw.publish("templates_index", bytes(PayloadList(items=log_templates_modified)))
+                log_templates_modified = []
+            last_time = start_time
+        if len(workload_model_logs) > 0:
+            await nw.publish("opnilog_workload_logs", bytes(PayloadList(items = workload_model_logs)))
+            logging.info(f"Published {len(workload_model_logs)} logs to be inferenced on by Workload Deep Learning model.")
+            workload_model_logs = []
+
+        logging.info(f"{len(logs_payload.items)} logs processed in {(time.time() - start_time)} second(s)")
+
+async def update_model(update_model_logs_queue, workload_template_miner):
+    '''
+    This function will process logs that were passed back by the Inferencing service. These log messages will be
+    added to the pretrained DRAIN model in addition to the predicted anomaly level as well.
+    '''
+    while True:
+        inferenced_logs = await update_model_logs_queue.get()
+        final_inferenced_logs = []
+        template_data = []
+        for log_data in inferenced_logs.items:
+            anomaly_level = log_data.anomaly_level
+            result = workload_template_miner.add_log_message(log_data.masked_log, log_data.log, anomaly_level)
+            log_data.template_matched = result["template_mined"]
+            log_data.template_cluster_id = result["cluster_id"]
+            if result["change_type"] == "cluster_created" or result["change_type"] == "cluster_template_changed":
+                template_data.append(Payload(log=result["sample_log"], template_matched=result["template_mined"], template_cluster_id=result["cluster_id"], _id="0{}".format(result["cluster_id"]), log_type=log_data.log_type))
+            final_inferenced_logs.append(log_data)
+        await nw.publish("inferenced_logs", bytes(PayloadList(items = final_inferenced_logs)))
+        await nw.publish("batch_processed_workload", "processed".encode())
```

Comment:
So essentially when the DRAIN workload model is being updated after the batched results are sent back from the inferencing service, it will send a signal to the batch_processed_workload Nats subject to indicate that the model has been updated and if it has been an hour since the last time the DRAIN model was saved into persistence, then that model will be saved again.